apps/emqx/test/emqx_persistent_session_ds_SUITE.erl

```diff
@@ -465,25 +474,104 @@ t_subscription_state_change(Config) ->
-            ),
+            %% Verify that QoS of subscription has been updated:
+            [#{packet_id := PI2, qos := ?QOS_2, topic := <<"t/1">>, payload := <<"2">>}] =
+                emqx_common_test_helpers:wait_publishes(1, 5_000),
             WaitGC(),
             #{subscriptions := Subs3, subscription_states := SStates3, streams := Streams3} = GetS(),
             ?assertEqual(Subs3, Subs2),
-            %% Old substate was deleted:
-            ?assertMatch([{SSID2, _}], maps:to_list(SStates3), Streams3)
+            %% Verify that the old substate (SSID1) was deleted:
+            ?assertMatch(
+                [{SSID2, _}],
+                maps:to_list(SStates3),
+                #{
+                    streams => Streams3,
+                    ssid1 => SSID1,
+                    ssid2 => SSID2,
+                    sub_states => SStates3
+                }
+            )
         end,
         []
     ).
 
+%% This testcase verifies the lifetimes of session's subscriptions to
+%% new stream events.
+t_new_stream_notifications(Config) ->
+    [Node1Spec | _] = ?config(node_specs, Config),
+    [Node1] = ?config(nodes, Config),
+    Port = get_mqtt_port(Node1, tcp),
+    ClientId = mk_clientid(?FUNCTION_NAME, sub),
+    TopicFilter1 = <<"foo/+">>,
+    TopicFilter2 = <<"bar">>,
+    ?check_trace(
+        #{timetrap => 30_000},
+        begin
+            %% Init:
+            Sub0 = start_client(#{port => Port, clientid => ClientId}),
+            {ok, _} = emqtt:connect(Sub0),
+            %% 1. Sessions should start watching streams when they
+            %% subscribe to the topics:
+            ?wait_async_action(
+                {ok, _, _} = emqtt:subscribe(Sub0, TopicFilter1, ?QOS_1),
+                #{?snk_kind := sessds_watch_streams, topic_filter := TopicFilter1}
+            ),
+            ?wait_async_action(
+                {ok, _, _} = emqtt:subscribe(Sub0, TopicFilter2, ?QOS_1),
+                #{?snk_kind := sessds_watch_streams, topic_filter := TopicFilter2}
+            ),
+            %% 2. Sessions should re-subscribe to the events after
+            %% reconnect:
+            emqtt:disconnect(Sub0),
+            {ok, SNKsub} = snabbkaffe:subscribe(
+                ?match_event(#{?snk_kind := sessds_watch_streams}),
+                2,
+                timer:seconds(10)
+            ),
+            Sub1 = start_client(#{port => Port, clientid => ClientId, clean_start => false}),
+            {ok, _} = emqtt:connect(Sub1),
+            %% Verify that both subscriptions have been renewed:
+            {ok, EventsAfterRestart} = snabbkaffe:receive_events(SNKsub),
+            ?assertMatch(
+                [<<"bar">>, <<"foo/+">>],
+                lists:sort(?projection(topic_filter, EventsAfterRestart))
+            ),
+            %% Verify that stream notifications are handled:
+            ?wait_async_action(
+                emqx_ds_new_streams:notify_new_stream(?PERSISTENT_MESSAGE_DB, <<"bar">>),
```

Comment:
Nit: this does not match `emqx_ds_new_streams:notify_new_stream/2`'s typespec: second argument should be a `emqx_ds:topic_filter()`, which is `list(binary() | '+' | '#' | '')`.